### PR TITLE
Add a testing principles section to dev handbook

### PR DESCRIPTION
**Pull request by agclever**
This is essentially moving some of the ideas I added in https://docs.google.com/document/d/1_TQrSeTocq-sROzI0fBKPZ-MRiMmoY1F0UsS9jt_a64/edit# to the dev-handbook.

I added a testing principles section and moved the section on testing deterministically from the `how` section. The rest of the stuff in the `how` section are ways of achieving the principles effectively as opposed to being more fundamental principles in my opinion.

The idea for this PR is two fold:
* Document these principles in a place that are easily accessible to existing and new clever engineers.
* Use this section as a template for a future post on the clever engineering blog.

### CODE REVIEW

**Comment by agclever** [comment]
cc: @Clever/engineers feel free to add your thoughts here. A lot of this stuff might be familiar to people who have already gone through the doc.

**Comment by schimmy** [comment]
I worry this page is a bit of a wall of text - I wonder if there is some way to pare the whole thing down or refactor it.

Otherwise, I like your additions! Maybe they belong in the "How Do You Write Tests" section, though?

File: testing.md [hunk 1]
```diff
@@ -34,6 +34,82 @@ A piece of code is not done until it's tested.
...
+and resetting it after execution. That is, the state before and after the test should be identical
+whether the test passes or fails. If your tests are independent, then running the tests in any order
+should not affect the outcome of the tests.
+
+A stronger version of independence could mean that there is no shared state between different tests.
+As an example, if the test involves talking to a database, then different tests should talk to
+different logical databases. Another example, if you are testing the behavior of a particular class,
+different tests should create their own instances of the class object. In this scenario, you could
+run all your tests in parallel keeping the outcome invariant. This becomes increasingly important
+as your code grows and your tests start to take longer . Retroactively making tests parallelizable
```
Comment (xavi-) [inline testing.md]: Nit.  Extra space after "start to take longer".

Comment (agclever) [inline testing.md]: Good catch. Fixed in 5272b08

**Comment by agclever** [comment]
@schimmy Agreed with this being a wall of text. One possible solution I can think of is partitioning this single doc into multiple ones linked to each other. Open to ideas here.

**Comment by agclever** [comment]
re: section, not super strong but I do feel that the `how` section contains a bunch of implementation details that feel distinct from the more root principals things in here.

File: testing.md [hunk 2]
```diff
@@ -34,6 +34,82 @@ A piece of code is not done until it's tested.
...
+of.
+
+### Prefer Readability and Ease of Code
+
+Tests should be written in a manner that prioritizes readability and ease of writing new tests and
+modifying existing ones. In particular, these should be prioritized over other typical concerns like
+abstraction and removing code duplication. If abstractions are needed, they should be created in a
+manner that preserves the logic in the test functions code without requiring many hops. Another way
+of saying this is that it should be apparent what the test is doing by just looking at the test
+code.
```
Comment (jonahkagan) [inline testing.md]: it's funny - this kind of sounds like a principle that should apply to all of the code we write. i think readability and maintainability are as important for all code, not just tests. why would tests have a different standard? this para doesn't really make a case why they should be different

File: testing.md [hunk 3]
```diff
@@ -34,6 +34,82 @@ A piece of code is not done until it's tested.
...
+
+Each test should be written in a manner that makes it independent of other tests. If there is any
+state needed for a test, the test should be responsible for getting to the state before running it
+and resetting it after execution. That is, the state before and after the test should be identical
+whether the test passes or fails. If your tests are independent, then running the tests in any order
+should not affect the outcome of the tests.
+
+A stronger version of independence could mean that there is no shared state between different tests.
+As an example, if the test involves talking to a database, then different tests should talk to
+different logical databases. Another example, if you are testing the behavior of a particular class,
```
Comment (jonahkagan) [inline testing.md]: setting up a different db for each test case seems costly in terms of time. that's the only real advantage to shared state that i can think of. how do you balance that concern?

File: testing.md [hunk 4]
```diff
@@ -34,6 +34,82 @@ A piece of code is not done until it's tested.
 
 So just write tests. Always.
 
+## Testing Principles
+
+Things you should keep in mind while writing tests for a codebase.
+
+### Write Complete Testing Suites
```
Comment (jonahkagan) [inline testing.md]: nit: the rest of the headers in this file only capitalize the first word

File: testing.md [hunk 5]
```diff
@@ -34,6 +34,82 @@ A piece of code is not done until it's tested.
...
+
+Testing suites should attempt to be exhaustive in verifying the code does what is expected. None of
+these tests should require any manual work, other than executing the initial command to run the
+tests.
+
+An exhaustive test suite implies that you should be confident of merging and deploying your changes
+once all your tests pass in your continuous integration environment. Any manual steps that are needed
+to verify code changes should be encoded as automated tests that run in your continuous integration
+environment.  As an example, if to test your code, you bring up an instance of your service and
```
Comment (jonahkagan) [inline testing.md]: continuous integration is introduced at the end of the file - feels a bit confusing to bring it up before its defined

File: testing.md [hunk 6]
```diff
@@ -34,6 +34,82 @@ A piece of code is not done until it's tested.
...
+manner that preserves the logic in the test functions code without requiring many hops. Another way
+of saying this is that it should be apparent what the test is doing by just looking at the test
+code.
+
+As an example, we had a testing library called which populated a bunch of fake data in the database
+specified by the test. A lot of our tests used with a single statement of the form
+`db_test_helpers.set_up_test_data`. This single function populated a bunch of state within the
+database including creating students, teachers, districts, apps. However, it also hid away all the
+details of the state being created. This meant that it became very hard to reason about tests or
+modify your tests if the current state was not appropriate for them.
```
Comment (jonahkagan) [inline testing.md]: in my experience, the main problem with having the shared fake data was that modifying the fake data impacted all of the dependent tests, not that the data was hidden. so i'm not sure this example is completely relevant to the point you're making here. is there another example you can think of?

**Comment by jonahkagan** [comment]
Nice additions!

File: testing.md (hunk 3 above)
Comment (agclever) [inline testing.md]: Notice this says logical databases not different servers. Correct me if I am wrong but setting up different logical dbs should be pretty cheap.

File: testing.md (hunk 4 above)
Comment (agclever) [inline testing.md]: Fixed in c8f9ba5

File: testing.md (hunk 3 above)
Comment (jonahkagan) [inline testing.md]: Setting them up is cheap, but populating them with data might not be.

File: testing.md (hunk 6 above)
Comment (agclever) [inline testing.md]: In my understanding fake data in the way it was used, had both problems. Theoretically it had could have been used to point to different DBs reducing the problem with dependent tests. Definitely open to better examples here.

File: testing.md [hunk 7]
```diff
@@ -34,23 +34,89 @@ A piece of code is not done until it's tested.
...
+can be hard while it's usually very little work to program them as such from the start.
 
-Concretely, this means you should avoid relying on factors like randomness, time, or external services in the structure of your tests. For instance, if you are testing a button that changes a piece of UI, you shouldn't wait for an arbitrary number of seconds to check that the UI refreshes, since the time that it takes to refresh is not deterministic. Instead, you can emit some sort of event when the UI is finished refreshing, and make the test listen for this event.
+## How do you write tests?
 
-Testing non-deterministic code can also be very difficult. If you can, avoid non-deterministic behavior in your code. If you must have it, try to isolate it so it can be deterministically mocked in tests.
+You'll probably want to use some sort of testing framework to help you write tests. Most languages support testing either with language constructs or with libraries. For example, we use [Mocha](http://visionmedia.github.io/mocha/) for node.js and [nose](https://nose.readthedocs.org/en/latest/) for Python.
```
Comment (nathanleiby) [inline testing.md]: could mention our Go practices as well

File: testing.md (hunk 2 above)
Comment (nathanleiby) [inline testing.md]: one area where you might have a different standard in non-test code: if performance is essential, you might choose a harder to read/use library if it had the best performance

but generally agreed that we should favor developer ease-of-use in all code!

**Comment by azylman** [comment]
@schimmy What's the state of this? If it's not relevant anymore, should we close it?

**Comment by schimmy** [comment]
hmm, up to @agclever - I haven't been working much on the eng blog as it seems we're doing ok on recruitment

**Comment by schimmy** [comment]
Assigning to you @agclever - I'm not too concerned about moving this forward for the blog, but you should perhaps merge it, it looks like a good improvement to me.

**Comment by agclever** [comment]
Will get back to this sometime this week with some changes since then.

**Comment by agclever** [comment]
Closing this for now. Should we revisited as we focus on shipping more safely.

**Pull request closed without merging**